### ISSUE TITLE
partial mappings for subgraph-isomorphism

### ISSUE THREAD

**Issue by boazpetersil** [networkx/networkx#3010]
Hi, 
This is a new-feature request or consultation (since i may try to implement this myself).

I wish to use GraphMatcher.subgraph_isomorphisms_iter() in order to retrieve candidates from a (graph) DB given a query. It would be very useful if in case the query does not appear in the database fully, i'd be able to find partial appearances of it. For example if the query consists of 5 nodes (with some connections between them), but only 4 of them appear in the DB, i would like to receive that partial mapping. 

I'm new to the field, but from my understanding, changing VF2 to this purpose shouldn't be too difficult. 
My thought was to use VF2 as it is, but on every backtrack save the current (partial) mapping and output it into a list. Because VF2  uses look ahead these partial mappings are usually smaller than they could be, so to solve this i thought i could iterate over the list and for each mapping i will run VF2 initlized to that mapping, and disable look-ahead. 

Does that make sense? Are there any potential problems i'm missing?

If not i will try to implement it, and if it works reasonably i'll be happy to contribute the code.

Thanks in advance,
Boaz

**Comment by chebee7i** [comment]
There's a bunch of partial mappings it encounters along the way. How are you defining which one that you care about?

In any event, if you rewrite the `match` method as an external function and change it slightly (so that it doesn't require that the mapping length is equal to G2), then I think you should be able to get it access to the mapping that is changing throughout the matching process.

**Comment by pckroon** [comment]
I'm (very) late to the party, but it sounds like you're looking for a largest common connected induced subgraph of a certain minimal size.
This is a non-trivial problem, unfortunately. Fortunately, some clever people already spend time on this: https://www.sciencedirect.com/science/article/pii/S0304397500002863
https://en.wikipedia.org/wiki/Maximum_common_induced_subgraph

I know from experience that it's not too hard to implement in networkx :) Essentially, you'll be looking for a largest clique of a certain minimum size in a modular product graph.

**Comment by jamestrimble** [comment]
> I'm (very) late to the party, but it sounds like you're looking for a largest common connected induced subgraph of a certain minimal size.

Would it be helpful if I had a go at implementing an algorithm for maximum common connected induced subgraph?

**Comment by dschult** [comment]
Yes, that sounds helpful. If I understand the comments above, this provides a way to find almost-isomorphisms. I don’t yet see the connection, but others have requested almost-isomorphisms.  ;)

**Comment by pckroon** [comment]
I think the algorithm by Koch (see the link I posted earlier) is still the current state-of-the-art. Also note that the ISMAGS isomorphism finder I implemented in networkx [1, 2] is also capable of finding the maximum common induced subgraph (not connected). I don't know how hard it'll be to add a `connected` keyword to that, but it may be worth looking at.

[1] https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.isomorphism.ISMAGS.largest_common_subgraph.html#networkx.algorithms.isomorphism.ISMAGS.largest_common_subgraph
[2] https://github.com/networkx/networkx/blob/master/networkx/algorithms/isomorphism/ismags.py#L544

**Comment by jamestrimble** [comment]
Thanks @pckroon . I’m sorry to say that after offering to help with an implementation, it now looks like I’m too busy in the near future to give it a try :-(

**Comment by pckroon** [comment]
With Hacktoberfest coming up I can probably make some time to implement either the algorithm in [1] or [2], unless someone else (@jamestrimble ?) calls dibs.

[1] https://link.springer.com/chapter/10.1007/978-3-319-94776-1_28
[2] https://www.sciencedirect.com/science/article/pii/S0166218X19301064#b20

**Comment by jamestrimble** [comment]
@pckroon I’ve managed to find some time to work on implementing a couple of maximum common subgraph algorithms (including a connected one), and I should hopefully be able to submit a PR with these in the next couple of days.

It would be great if you could implement the _maximal_ connected subgraph algorithms you cite.

**Comment by pckroon** [comment]
Cool! :)
I'll try the BC-ENUM algorithm probably (first reference), but I won't make any promises.

**Comment by jamestrimble** [comment]
@pckroon It occurred to me that it might be possible to repurpose the McSplit algorithm to solve the maximal common connected induced subgraph enumeration problem. I've made a quick attempt at https://github.com/jamestrimble/mcsplit-maximal-common-subgraph . I tried it out on a few random graphs and it gives the same number of subgraphs as veluca93's BC-ENUM program, which is encouraging! The performance seems reasonable, although I haven't done proper experiments.

Perhaps it might be helpful to have this alternative implementation for unit testing if nothing else?

**Comment by pckroon** [comment]
Thanks, looks interesting :-)
I'll take a more careful look once I start the implementation, there's a IRL corona-induced mess here I have to deal with at the moment....

**Comment by jamestrimble** [comment]
@pckroon I hope things are going okay with the real-life problems you mentioned in your previous post. I have been looking a bit more into Koch's algorithm, as it's relevant to my PhD work. My non-NetworkX implementations are at https://github.com/jamestrimble/koch-mccis. Luca Versari has two very nice C++ versions using an implicit product graph in his batchelor's thesis, with the code in the appendix: https://www.lucaversari.it/bachelors/thesis.pdf.

I'm just passing this info along in case it might be useful to you. I don't have time to do polished NetworkX implementations of any of these maximal CCIS algorithms myself at the moment.

**Comment by pckroon** [comment]
Hey @jamestrimble thanks, for checking up on me. Fortunately the mess was purely organisational (they tightened Corona measures 4 days before my PhD defense...), but it all worked out fine.
I haven't really found the time yet to try and implement the promised algorithm. Still about 2 weeks left for hacktoberfest though. And if I don't make that it'll come a bit after.
I'll first try to implement the algorithm from just the paper, but once that fails I'll be sure to check out your info :)

**Comment by jamestrimble** [comment]
@pckroon Congratulations!!!
